rfd/0030-upgrade-system.md

```diff
@@ -0,0 +1,610 @@
+have heartbeated since the release time should be considered valid upgrade targets (so that the
+version information in the backend matches the post-upgrade state of the world).
+
+Regardless of the lease strategy, leases should include some basic information, including:
+
+- Name of the controller that created the lease.
+- Est. number of servers targeted for upgrade (this might be the number of servers which
+match the `server_selector` and not necessarily reflective of the number of servers that
+will actually get upgraded).
+- Version being targeted.
+
+
+#### Upgrade UX
+
+At the cluster-level, upgrade flows should roughly fall into two camps.  Manual and automatic,
+where manual upgrades require some kind of triggering from a user.  Other than needing a simple
+manual trigger, the UX around these two modes of operation should be identical.  I think the
+best way to do this is to have the process of applying policies to available releases produce
+a compiled "target state" that is independent of the resources used to calculate it (i.e. the
+upgrade policies and release manifests).
+
+By keeping the target state indipendent of the resources used to calculate it, we are able to
+store it as an indipendent resource.  Upgrade controllers will attempt to converge toward the stored
+target state, meaning that they will not be directly affected by new releases becoming available
+or upgrade policies being changed.  If the cluster is configured for automatic upgrades, then
+periodic operations in the auth server can recompile the target state.  If the cluster is configured
+for manual upgrades, the user can preview and apply new compilations of the target state via `tctl`.
+
+This should allow us to implement a manual upgrade flow that looks something like this:
+
+```
+$ tctl upgrade
+<diff of target state>
+Proceed: y/n
+```
+
+Or, if we want to allow for manual modification of target state, something like:
+
+```
+$ tctl upgrade plan > target-state.yaml
+# review the plan & make changes
+
+$ tctl upgrade apply -f target-state.yaml
+<diff>
+Proceed: y/n
+```
+
+### Technical Notes
+
+
+#### Crypto
+
+We aren't in full control of the cryptography at play since we're integrating user scripts and
+third-party installation mechanisms, but where possible we will be opinionated.  Where applicable
+we will select algorithms that are relatively modern and targeted to meet or exceede a 128-bit
+security level.
+
+##### Package Signing
+
+This is generally covered by existing package distribution solutions.  I'm leaning toward `minisign`
+being the distributor-agnostic signing mechanism of choice.  Its dead simple, produces signatures
+compatible with OpenBSD's `signify`, and the one piece of "bloat" it does have, trusted comments,
+is actually useful to us.
+
+##### Package Checksums
+
+256-bit `BLAKE2` will be used for checksums (compatible with the `b2sum` utility from GNU coreutils).
```

Comment:
I don't believe blake2 is covered by the CMP program, so this may not be a good choice for FIPS builds using fips validated crypto modules.

Reply:
Good point.  I hadn't considered fips compliance here.  `SHA-3` it is then.